cassandra-protocol/src/frame.rs

```diff
@@ -255,16 +255,59 @@ impl Envelope {
         v.extend_from_slice(&self.stream_id.to_be_bytes());
         v.push(opcode_byte);
 
+        let mut flags_buffer = vec![];
+
+        if self.flags.contains(Flags::TRACING) && self.direction == Direction::Response {
+            let mut tracing_id = self
+                .tracing_id
+                .ok_or_else(|| {
+                    error::Error::Io(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        "Tracing flag was set but Envelope has no tracing_id",
+                    ))
+                })?
+                .into_bytes()
+                .to_vec();
+
+            flags_buffer.append(&mut tracing_id);
+        };
+
+        if self.flags.contains(Flags::WARNING) && self.direction == Direction::Response {
+            let mut warnings_buffer = vec![];
+
+            for warning in &self.warnings {
+                let warning_len = warning.len() as i16;
+                warnings_buffer.extend_from_slice(&warning_len.to_be_bytes());
+                warnings_buffer.append(&mut warning.as_bytes().to_vec());
+            }
+
+            let warnings_len = self.warnings.len() as i16;
+            flags_buffer.extend_from_slice(&warnings_len.to_be_bytes());
+            flags_buffer.append(&mut warnings_buffer);
+        }
+
         if is_compressed {
-            let mut encoded_body = compressor.encode(&self.body)?;
+            // avoid concating the slices if there is nothing in flags_buffer
+            let encoded_body = if flags_buffer.is_empty() {
+                compressor.encode(&self.body)?
+            } else {
+                compressor.encode(&[flags_buffer.as_slice(), &self.body].concat())?
```

Comment:
`concat` should not be used as it creates a temporary vector and copies all the slices into it.